Draft: RFC 6761 handling for invalid and *.localhost

gbr-mendes opened the pull request:
This is a draft PR for [issue #118569](https://github.com/dotnet/runtime/issues/118569).

### Motivation
RFC 6761 defines special-use domain names:
- `invalid` and `*.invalid` must always return NXDOMAIN
- `localhost` and `*.localhost` must always resolve to the loopback address

Currently, .NET relies fully on the OS resolver, which may allow `invalid` to resolve if defined in `/etc/hosts` or DNS, and may not handle `*.localhost` consistently across platforms.

### Proposed approach
- Intercept `invalid` and `*.invalid` before OS resolution → throw `SocketException` with `HostNotFound` (simulate NXDOMAIN).
- Intercept `localhost` and `*.localhost` before OS resolution → return `IPAddress.Loopback` and `IPAddress.IPv6Loopback`.
- Otherwise, fallback to the platform resolver as today.

### Status
- Draft PR to validate the idea/approach with maintainers.
- No tests included yet.
- Once approach is confirmed, I’ll proceed with the real implementation, unit tests and CI validation.

gbr-mendes committed:
Draft: RFC 6761 handling for invalid and *.localhost

rzikm reviewed (commented):
I think the approach is fine, but needs a bit of improvement and testing.

wfurt commented on src/libraries/System.Net.NameResolution/src/System/Net/Dns.cs:
```diff
@@ -426,6 +426,19 @@ private static object GetHostEntryOrAddressesCore(string hostName, bool justAddr
+                throw new SocketException((int)SocketError.HostNotFound);
+            }
+
+            if (hostName.Equals("localhost", StringComparison.OrdinalIgnoreCase) ||
+                hostName.EndsWith(".localhost", StringComparison.OrdinalIgnoreCase))
+            {
+                IPAddress[] loopbacks = new IPAddress[] { IPAddress.Loopback, IPAddress.IPv6Loopback };
+                return justAddresses ? (object)loopbacks : new IPHostEntry { AddressList = loopbacks, HostName = hostName, Aliases = Array.Empty<string>() };
+            }
```
I'm also wondering if one could pre-allocate immutable answer is this should probably not change through the process life.

wfurt commented on src/libraries/System.Net.NameResolution/src/System/Net/Dns.cs:
```diff
@@ -426,6 +426,19 @@ private static object GetHostEntryOrAddressesCore(string hostName, bool justAddr
                 return resultOnFailure;
             }
 
+            if (hostName.Equals("invalid", StringComparison.OrdinalIgnoreCase) ||
```
may be worth of diag log as well.

filipnavara commented on src/libraries/System.Net.NameResolution/src/System/Net/Dns.cs:
```diff
@@ -426,6 +426,19 @@ private static object GetHostEntryOrAddressesCore(string hostName, bool justAddr
                 return resultOnFailure;
             }
 
+            if (hostName.Equals("invalid", StringComparison.OrdinalIgnoreCase) ||
+                hostName.EndsWith(".invalid", StringComparison.OrdinalIgnoreCase))
```
That doesn't sound like a correct comparison. It only checks for the `.` character and not the rest of the string, so it would also match `foo.xxvalid` with reservedName == `invalid`.

filipnavara replied:
You could probably do something like `name.EndsWith(reservedName, StringComparison.OrdinalIgnoreCase) && (name.Length == reservedName.Length || (name.Length > reservedName.Length && name[name.Length - reservedName.Length - 1] == '.'))`. It's not very readable though...

rzikm replied:
right, I messed it, it was supposed to be as you wrote.

As for readability, that's what the descriptive name of the helper function is for :)

gbr-mendes committed:
Draft: Refine RFC 6761 handling for invalid and *.localhost

- Consolidated reserved name checks into MatchesReservedName helper
- Return only supported IP versions in GetLoopbacksForAddressFamily
- Added diagnostic logging for special-use domains
- Adjusted handling to respect AddressFamily and OS IPv4/IPv6 support
- Special-name handling now runs between BeforeResolution and AfterResolution

gbr-mendes commented:
> @gbr-mendes please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @dotnet-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @dotnet-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @dotnet-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement

@dotnet-policy-service agree

rzikm reviewed (commented):
Sorry for ignoring this, it slipped my attention.

@wfurt do you have any thoughts on how to check for IPv4/IPv6 support reliably?
